src/browser/sse.ts

```diff
@@ -0,0 +1,764 @@
+    [key: string]: unknown
+  },
+  Params extends PathParams,
+> = ResponseResolver<ServerSentEventResolverExtras<EventMap, Params>, any, any>
+
+export type ServerSentEventRequestHandler = <
+  EventMap extends {
+    message?: unknown
+    [key: string]: unknown
+  },
+  Params extends PathParams<keyof Params> = PathParams,
+  RequestPath extends Path = Path,
+>(
+  path: RequestPath,
+  resolver: ServerSentEventResolver<EventMap, Params>,
+) => HttpHandler
+
+/**
+ * Request handler for Server-Sent Events (SSE).
+ *
+ * @example
+ * sse('http://localhost:4321', ({ source }) => {
+ *   source.send({ data: 'hello world' })
+ * })
+ *
+ * @see {@link https://mswjs.io/docs/api/sse `sse()` API reference}
+ */
+export const sse: ServerSentEventRequestHandler = (path, resolver) => {
+  return new ServerSentEventHandler(path, resolver)
+}
+
+class ServerSentEventHandler<
+  EventMap extends {
+    message?: unknown
+    [key: string]: unknown
+  },
+> extends HttpHandler {
+  constructor(path: Path, resolver: ServerSentEventResolver<EventMap, any>) {
+    invariant(
+      typeof EventSource !== 'undefined',
+      'Failed to construct a Server-Sent Event handler for path "%s": your environment does not support the EventSource API',
+      path,
+    )
+
+    super('GET', path, (info) => {
+      const stream = new ReadableStream({
+        start(controller) {
+          const client = new ServerSentEventClient<EventMap>({
+            controller,
+          })
+          const server = new ServerSentEventServer({
+            request: info.request,
+            client,
+          })
+
+          resolver({
+            ...info,
+            client,
+            server,
+          })
+        },
+      })
+
+      return new Response(stream, {
+        headers: {
+          'content-type': 'text/event-stream',
+          'cache-control': 'no-cache',
+          connection: 'keep-alive',
+        },
+      })
+    })
+  }
+
+  predicate(args: {
+    request: Request
+    parsedResult: HttpRequestParsedResult
+  }): boolean {
+    // Ignore requests that are not SSE.
+    if (args.request.headers.get('accept') !== 'text/event-stream') {
+      return false
+    }
+
+    return super.predicate(args)
+  }
+}
+
+class ServerSentEventClient<
+  EventMap extends {
+    message?: unknown
+    [key: string]: unknown
+  },
+> {
+  private encoder: TextEncoder
+  private controller: ReadableStreamDefaultController
+
+  constructor(args: { controller: ReadableStreamDefaultController }) {
+    this.encoder = new TextEncoder()
+    this.controller = args.controller
+  }
+
+  /**
+   * Sends the given payload to the intercepted `EventSource`.
+   */
+  public send<EventType extends keyof EventMap = 'message'>(
+    payload:
+      | {
+          id?: string
+          event?: 'message'
+          data?: EventMap['message']
+          retry?: never
+        }
+      | {
+          id?: string
+          event?: EventType
+          data?: EventMap[EventType]
```

Comment:
Should this be a valid call?
```ts
sse<{ custom: string; }>('/stream', ({ client }) => {
  client.send({ type: "custom" })
})
```

IMHO, it shouldn't be because the user forgot to pass `data`. But currently it is valid because `data` is always optional here

Reply:
It shouldn't. Events without data must be allowed but if you've described an event having a data type in your events map, then you have to pass `data` (otherwise use `{ custom: never }`).